OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignal.java

```diff
@@ -1985,18 +2074,51 @@ public void run() {
       });
    }
 
-   // Called when receiving FCM/ADM message after it has been displayed.
-   // Or right when it is received if it is a silent one
-   //   If a NotificationExtenderService is present in the developers app this will not fire for silent notifications.
-   static void handleNotificationReceived(JSONArray data, boolean displayed, boolean fromAlert) {
-      OSNotificationOpenResult openResult = generateOsNotificationOpenResult(data, displayed, fromAlert);
-      if(trackFirebaseAnalytics != null && getFirebaseAnalyticsEnabled())
-         trackFirebaseAnalytics.trackReceivedEvent(openResult);
+   /**
+    * Called when receiving FCM/ADM message after it has been displayed.
+    * Or right when it is received if it is a silent one
+    *   If a NotificationExtenderService is present in the developers app this will not fire for silent notifications.
+    */
+   static void handleNotificationReceived(OSNotificationGenerationJob notifJob, boolean displayed) {
+      try {
+         JSONObject jsonObject = new JSONObject(notifJob.jsonPayload.toString());
+         jsonObject.put(BUNDLE_KEY_ANDROID_NOTIFICATION_ID, notifJob.getAndroidId());
+
+         OSNotificationOpenResult openResult = generateOsNotificationOpenResult(newJsonArray(jsonObject), displayed);
+         if (trackFirebaseAnalytics != null && getFirebaseAnalyticsEnabled())
+            trackFirebaseAnalytics.trackReceivedEvent(openResult);
+
+      } catch (JSONException e) {
+         e.printStackTrace();
+      }
+   }
+
+   /**
+    * A few checks before before firing the normal expected scenario where a ExtNotificationWillShowInForegroundHandler
+    *  and AppNotificationWillShowInForegroundHandler are set
+    * <br/><br/>
+    * @see OneSignal.ExtNotificationWillShowInForegroundHandler
+    * @see OSNotificationWillShowInForegroundManager.ExtNotificationWillShowInForegroundWorker
+    * @see OneSignal.AppNotificationWillShowInForegroundHandler
+    * @see OSNotificationWillShowInForegroundManager.AppNotificationWillShowInForegroundWorker
+    */
+   static void fireNotificationWillShowInForegroundHandlers(OSNotificationGenerationJob notifJob) {
+      // No ExtNotificationWillShowInForegroundHandler or AppNotificationWillShowInForegroundHandler was setup, show the notification
+      if (extNotificationWillShowInForegroundHandler == null && appNotificationWillShowInForegroundHandler == null) {
+         OneSignal.onesignalLog(OneSignal.LOG_LEVEL.INFO, "extNotificationWillShowInForegroundHandler and appNotificationWillShowInForegroundHandler are both null, moving on to showing OS notification id: " + notifJob.getApiNotificationId());
+         GenerateNotification.fromJsonPayload(notifJob);
```

Comment:
`GenerateNotification.fromJsonPayload(notifJob);` is called from within this `fireNotificationWillShowInForegroundHandlers`. Would be more clear if `fireNotificationWillShowInForegroundHandlers` would return a `boolean` that it did create an event or not then call `GenerateNotification.fromJsonPayload(notifJob);` outside of this method.

Reply:
Got it, sounds good!